### DISCUSSION TITLE
Custom headers overriding doesnt work when a header is deleted

### DISCUSSION THREAD

**Discussion by vacavaca** [vercel/next.js#46292] [Ideas]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```
    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:08:47 PST 2022; root:xnu-8792.61.2~4/RELEASE_X86_64
    Binaries:
      Node: 18.0.0
      npm: 8.6.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.1.7-canary.13
      eslint-config-next: 13.1.6
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue

https://codesandbox.io/p/sandbox/weathered-cloud-9srp6h?file=%2Fpages%2Findex.tsx&selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A5%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A5%7D%5D

### To Reproduce

next.config.js:
```
async headers() {
    return [
      {
        source: '/:path*',
        headers: [
          {
            key: 'X-Test',
            value: 'Test value'
          }
        ]
      },
      {
        source: '/sub/:path*',
        headers: [
          {
            key: 'X-Another',
            value: 'Another header'
          }
        ]
      }
    ]
  }
```

### Describe the Bug

Deleting a custom header is not possible by overriding headers array in next.config.js:

In the example above, pages on the /sub path will also get "X-Test" header, regardless of the setting in the configuration.

### Expected Behavior

The entire headers array from previous entries must be overridden.

If this is an intended behavior, then please consider the following use-case:

It's not uncommon to intentionally weaken security requirements for some pages in the app. For example, when the app needs to allow access to some of the pages from an `iframe` from another site by unsetting 'X-Frame-Options' header from some of the routes.

When routing structure is simple, it is straightforward to put such pages on completely separate sub paths like example.com/inside-iframe/.... and to define two separate sets of custom header for different parts of the app.

But when we consider a more real-world example, there might be a lot of other logic and features that require a more complicated routing structure with lots of dynamic routes and different relations between them, their sub paths, sub-sub paths etc.

It's possible to make a workaround solution, for example by manually modifying headers on a reverse-proxy or by using a middleware, but it will get messy fast, especially in the case with mixed dynamic (with a parameter) and static pages under the same path.

I can't come up with a decent solution for such situation:

```
/sub/:param   - remove one of security headers
/sub/page  	- keep the header
/.... (all other routes) - keep the header
```

This requires me to remember to modify my reverse-proxy / middleware each time when a new page is created under /sub path. That easily leads to some pretty serious security issues.

I could set the header to an empty string in such a case but this is also not an ideal solution.

Is there a better workaround? Or maybe this is indeed a bug and will be fixed in the future versions?

<sub>[NEXT-556](https://linear.app/vercel/issue/NEXT-556/custom-headers-overriding-doesnt-work-when-a-header-is-deleted)</sub>

**Comment by stmtk1** [comment]
You write '/:id*' source value. This is not collect format. Collect format is '/:id'.
If you remove `*` after `id`, custom header will work.

**Comment by vacavaca** [comment]
Could you please clarify a bit? I've tested this exact configuration (in the code-sandbox), and it worked as I described here: a new custom header is added, but the old one is not removed.

The format is from the [docs](https://nextjs.org/docs/api-reference/next.config.js/headers), but it lacks details a bit, so I could've got it wrong.

**Comment by stmtk1** [comment]
Oh, I didn't know wild card matching. My comment is not fact. So please ignore my comment.
